### PR TITLE
feat(agw): protobuf/json file based config and loader

### PR DESCRIPTION
**Pull request by markjen**
#8832 

This is a stack on top of #9067 and #9038. To review this change before the others land and this PR is rebased, see [046675a4bda23450a1f01fdcd478c4b58bb4c7ba](https://github.com/magma/magma/commit/046675a4bda23450a1f01fdcd478c4b58bb4c7ba)

## Summary

- add config path flag
- config loader package
- support comments in config json
- move all grpc targets to config

## Test Plan

```
$ go test -cover ./accessd/config/...
ok  	github.com/magma/magma/accessd/config	(cached)	coverage: 98.0% of statements
ok  	github.com/magma/magma/accessd/config/internal/grpcutil	(cached)	coverage: 100.0% of statements
```

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: src/go/accessd/config/config.go [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+const (
+	ipv4Scheme = "ipv4"
+	ipv6Scheme = "ipv6"
+)
+
+// ParseTarget takes a target in string form and returns a resolved Target.
+// Extends functionality in grpc/internal/grpcutil.ParseTarget to support ipv4
+// and ipv6 schemes.
+func ParseTarget(target string) resolver.Target {
```
Comment (electronjoe) [inline src/go/accessd/config/config.go]: nit: I could see this utility function eventually being used across go daemons - so present location in accessd config dir / module feels short term.

File: src/go/accessd/config/config.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+			break
+		}
+		in = in[idx+2:]
+	}
+	return filtered
+}
+
+func filterJSONComments(in string) string {
+	var filtered []string
+	for _, line := range strings.Split(filterCStyleComments(in), "\n") {
```
Comment (electronjoe) [inline src/go/accessd/config/config.go]: Are our comments in the JSON configs not to spec? Should they not be there? Is there any vulnerability to the filters mis-firing on valid fields in the configuration data (e.g. strings containing `//` or `/*` etc as data fields)? I get nervous when we implement our own parser libs (which this kind of is), as it's so hard to get right right.

Comment (electronjoe) [reply src/go/accessd/config/config.go]: Upshot - is this another tech debt example, and if so should we file a GH Issue to remove these comments and tag issue here for removal of this code?

File: src/go/accessd/config/config.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+func NewConfigManager(configFilePath string) (*ConfigManager, error) {
+	config, ok := proto.Clone(defaultConfig).(*mconfig.AccessD)
+	if !ok {
+		panic("clone of defaultConfig not *mconfig.AccessD")
+	}
+	if _, err := os.Stat(configFilePath); err == nil {
+		loaded, err := loadMConfigProto(
+			ioutil.ReadFile, protojson.Unmarshal, configFilePath)
+		if err != nil {
+			return nil, err
```
Comment (electronjoe) [inline src/go/accessd/config/config.go]: So to be clear, the behavior here appears to be:
- Config file must exist and be parsed successfully
  - otherwise return error and nil config
- Upon successful read and parse of config, merge with defaults

Behavior is a little nuanced and requires a careful read - might be worth stating something like this on NewConfigManager function documentation?

File: src/go/accessd/config/config_test.go [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+	t.Parallel()
+
+	tests := []struct {
+		target string
+		parsed resolver.Target
+	}{
+		{},
+		{
+			target: "unix:///tmp/sctpd_downstream.sock",
+			parsed: resolver.Target{
```
Comment (electronjoe) [inline src/go/accessd/config/config_test.go]: nit: We used `want` for a data driven testing field that was singularly being compared in the test - to make the for loop easy to read below: `assert.Equal(t, test.want, got)`

File: src/go/accessd/config/config_test.go [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+				Scheme:    "ipv6",
+				Authority: "",
+				Endpoint:  "[::]:1234",
+			},
+		},
+	}
+
+	for _, test := range tests {
+		got := ParseTarget(test.target)
+		assert.Equal(t, test.parsed, got)
```
Comment (electronjoe) [inline src/go/accessd/config/config_test.go]: nit: when this asserts it will not present the reviewer with the input? Though it can be inferred from test.want I suppose. [Identify the Input](https://github.com/golang/go/issues/34384).

File: src/go/accessd/config/config_test.go (hunk 4 above)
Comment (electronjoe) [reply src/go/accessd/config/config_test.go]: I do see you did this in table driven tests below (just not here)

File: src/go/accessd/config/internal/grpcutil/target_test.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+			t.Errorf("ParseTarget(%q, false) = %+v, want %+v", str, got, test)
+		}
+		got = ParseTarget(str, true)
+		if got != test {
+			t.Errorf("ParseTarget(%q, true) = %+v, want %+v", str, got, test)
+		}
+	}
+}
+
+func TestParseTargetString(t *testing.T) {
```
Comment (electronjoe) [inline src/go/accessd/config/internal/grpcutil/target_test.go]: Oof! Lots of permutations to test here.  I guess no function like this exists in a well tested library?

File: src/go/accessd/config/testdata/accessd_mconfig.json [hunk 7]
```diff
@@ -0,0 +1,16 @@
+// comment outside of json
+{
```
Comment (electronjoe) [inline src/go/accessd/config/testdata/accessd_mconfig.json]: Lots to say here, which I learned last week in the config debug on master.

The below is intended to describe how the Magma AGW presently operates, per my understanding from discussions with @ardzoht and review of [magma dev documentation](https://docs.magmacore.org/docs/lte/dev_notes).

## mconfig files

As part of Magma install, a single [gateway.mconfig](https://sourcegraph.com/github.com/magma/magma/-/blob/lte/gateway/configs/gateway.mconfig?subtree=true) is placed at `/etc/magma/gateway.mconfig`. Note that it has a special shape and expects an initial JSON key of `configs_by_key` - then an array of per-daemon protobuf representations with a `"@type"` annotation.

If you were going to use mconfig for your configuration, you would modify this static configuration...

But - big but - as part of check-in with an `Orc8r` instance, `Orc8r` will push new configs to the AGW. This new `mconfig` is then stored ephemerally to `/var/opt/magma/configs/gateway.mconfig` and once present completely disregards any contents of your static `/etc/magma/gateway.mconfig`.

So if you are going to go the `mconfig` route, you have to wire this configuration change in through `NMS UI` and `Orc8r config push` all the way to the mconfg write to `/var/`.. Otherwise the first Orc8r push will over-write your values.

There is an alternative...

## YAML Service Configs

So for local AGW service behavior configuration, the intended path today is to write a YAML service config to `/etc/magma/<service_name>.yml`.

I am unclear whether any pre-defined schema exists for these files, I think it does not. So they are totally at-will w.r.t. additions / expectation of content.  You would then read that configuration YAML and operate on it at runtime, for any fields it defines.

**Review by electronjoe — APPROVED** [review]

File: src/go/accessd/config/config.go (hunk 2 above)
Comment (markjen) [reply src/go/accessd/config/config.go]: We want to support comments in JSON files for documentation purposes. Other popular progarms (e.g. VSCode) use a settings.json file and support comments.

(Also, a stated reason behind wanting to use YAML as a config language is because it supports comments. With this addition, we are one step closer to eliminating YAML :) )

Background:

1. JSON spec does not support comments (they are explicitly disallowed, https://en.wikipedia.org/wiki/JSON#cite_note-DouglasCrockfordComments-24). A pre-processor is recommended/suggested
2. Golang's built in json parsing does not support comments. This is by design/follows their conventions: they are very strict about adherence to official specs.
3. AFAICT, no well tested pre-processor exists. I found a few, but they don't seem to have good tests, e.g.

https://github.com/wuxiangzhou2010/jsonuncommenter/blob/master/uncommenter_test.go
https://github.com/muhammadmuzzammil1998/jsonc/blob/master/jsonc_test.go

4. Overall, this pre-processor is ~30 lines of code easily testable. Not quite a "leftpad" situation, but pretty close :)

File: lte/protos/mconfig/mconfigs.proto [hunk 8]
```diff
@@ -468,3 +468,23 @@ message GatewayDNSConfigRecordsItems {
...
+    ERROR = 4;
+  }
+  LogLevel log_level = 1;
+
+  string sctpd_downstream_service_target = 2;
+  string sctpd_upstream_service_target = 3;
+  string mme_sctpd_downstream_service_target = 4;
+  string mme_sctpd_upstream_service_target = 5;
+
+}
```
Comment (uri200) [inline lte/protos/mconfig/mconfigs.proto:489]: we already have a service called `accessd` on orc8r. 

Is this service related to that one? If not, maybe it would be good to find a better name?

File: src/go/accessd/config/config.go (hunk 3 above)
Comment (markjen) [reply src/go/accessd/config/config.go]: Will add documentation (and this is also covered in the test cases)

The current behavior is:
1. If no file exists, you get default
2. If a file exists, it must be readable and parsable (or error)
3. Once read, it overrides non-default/zero values

File: src/go/accessd/config/internal/grpcutil/target_test.go (hunk 6 above)
Comment (markjen) [reply src/go/accessd/config/internal/grpcutil/target_test.go]: this is extracted from the grpc lib :)

File: src/go/accessd/config/config.go [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+		}
+		proto.Merge(config, loaded)
+	}
+	return &ConfigManager{config: config}, nil
+}
+
+// Config returns the parsed config contained in ConfigManager.
+func (c *ConfigManager) Config() *mconfig.AccessD {
+	return c.config
+}
```
Comment (uri200) [inline src/go/accessd/config/config.go]: this maybe a naive question, but it looks like you are loading mconfig file

Why can't we use `GetServiceConfig` we use in the other gateways [here](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Auri200%2Fmagma.git&branch=state_service_add_print_grpc&file=feg%2Fgateway%2Fservices%2Fs8_proxy%2Fservicers%2Fconfig.go&editor=JetBrains&version=v1.2.1&utm_product_name=IntelliJ+IDEA&utm_product_version=2021.1.3&start_row=39&start_col=29&end_row=39&end_col=29). 



Note that our mcofigs values come in string format and sometimes they need some massaging like [here](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Auri200%2Fmagma.git&branch=state_service_add_print_grpc&file=feg%2Fgateway%2Fservices%2Fs8_proxy%2Fservicers%2Fconfig.go&editor=JetBrains&version=v1.2.1&utm_product_name=IntelliJ+IDEA&utm_product_version=2021.1.3&start_row=47&start_col=31&end_row=47&end_col=31)

**Comment by uri200** [comment]
Hey @markjen per our conversation, I leave here a rough description of how we build a service using the orc8r framework. I will be using s8_proxy feg service as an example

We define the new service the [main](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Auri200%2Fmagma.git&branch=state_service_add_print_grpc&file=feg%2Fgateway%2Fservices%2Fs8_proxy%2Fs8_proxy%2Fmain.go&editor=JetBrains&version=v1.2.1&utm_product_name=IntelliJ+IDEA&utm_product_version=2021.1.3&start_row=30&start_col=4&end_row=30&end_col=4) file of our service:

1- Create a new grpc service using `NewServiceWithOptions`: [here](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Auri200%2Fmagma.git&branch=state_service_add_print_grpc&file=feg%2Fgateway%2Fservices%2Fs8_proxy%2Fs8_proxy%2Fmain.go&editor=JetBrains&version=v1.2.1&utm_product_name=IntelliJ+IDEA&utm_product_version=2021.1.3&start_row=37&start_col=41&end_row=37&end_col=41) we create a generic `grpc.Server` and preconfigure grpc options like `KeepaliveParams`. We also add`moduleName` and `serviceName` which will be used by controlProxy and Magmad to route those request (the service name matches [service_registry.yml](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Auri200%2Fmagma.git&branch=state_service_add_print_grpc&file=feg%2Fgateway%2Fconfigs%2Fservice_registry.yml&editor=JetBrains&version=v1.2.1&utm_product_name=IntelliJ+IDEA&utm_product_version=2021.1.3&start_row=26&start_col=0&end_row=27&end_col=0) on the feg)

2- Parse configuration for that specific service: all services have a [GetConfig](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Auri200%2Fmagma.git&branch=state_service_add_print_grpc&file=feg%2Fgateway%2Fservices%2Fs8_proxy%2Fs8_proxy%2Fmain.go&editor=JetBrains&version=v1.2.1&utm_product_name=IntelliJ+IDEA&utm_product_version=2021.1.3&start_row=37&start_col=41&end_row=37&end_col=41) function which does the task of grabbing the gateway.mconfig and parsing the specific fields for that service. In case of `s8_proxy` we identify the values inside mconfig using `S8ProxyServiceName` [here](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Auri200%2Fmagma.git&branch=state_service_add_print_grpc&file=feg%2Fgateway%2Fservices%2Fs8_proxy%2Fservicers%2Fconfig.go&editor=JetBrains&version=v1.2.1&utm_product_name=IntelliJ+IDEA&utm_product_version=2021.1.3&start_row=39&start_col=7&end_row=39&end_col=7)

3- Creation of the implementation of the grpc service: here we create an implementation of `S8ProxyServer` which includes all the grpc calls. That is the actual [s8_proxy](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Auri200%2Fmagma.git&branch=state_service_add_print_grpc&file=feg%2Fgateway%2Fservices%2Fs8_proxy%2Fs8_proxy%2Fmain.go&editor=JetBrains&version=v1.2.1&utm_product_name=IntelliJ+IDEA&utm_product_version=2021.1.3&start_row=40&start_col=33&end_row=40&end_col=33) logic 

4- Register the implementation of the service into the grpc server

5- run the server


______________________________________________________
An example of how we parse a config can be found at s6a_proxy [here](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Auri200%2Fmagma.git&branch=state_service_add_print_grpc&file=feg%2Fgateway%2Fservices%2Fs6a_proxy%2Fservicers%2Fconfig.go&editor=JetBrains&version=v1.2.1&utm_product_name=IntelliJ+IDEA&utm_product_version=2021.1.3&start_row=47&start_col=17&end_row=47&end_col=17)

There we parse a specific key for that service on mconfig file. Once it is parsed we manually put each field on the servcie config file (in case of s6a_proxy this is `S6aProxyConfig`). 

In case mconfig file is not found, we build a default config file using local envars [here](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Auri200%2Fmagma.git&branch=state_service_add_print_grpc&file=feg%2Fgateway%2Fservices%2Fs6a_proxy%2Fservicers%2Fconfig.go&editor=JetBrains&version=v1.2.1&utm_product_name=IntelliJ+IDEA&utm_product_version=2021.1.3&start_row=59&start_col=13&end_row=59&end_col=13)

**Comment by jwagantall** [comment]
RestartCWAGTests

File: src/go/accessd/config/config.go (hunk 9 above)
Comment (markjen) [reply src/go/accessd/config/config.go]: trying to see if there's a easier way to do configs. we will follow up here in a sprint or two as part of #8832 with a potentially bigger redesign of config (also with a path towards unifying a new config strategy with config across the rest of our services/codebase)

**Comment by markjen** [comment]
> Hey @markjen per our conversation, I leave here a rough description of how we build a service using the orc8r framework. I will be using s8_proxy feg service as an example
> 
> We define the new service the [main](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Auri200%2Fmagma.git&branch=state_service_add_print_grpc&file=feg%2Fgateway%2Fservices%2Fs8_proxy%2Fs8_proxy%2Fmain.go&editor=JetBrains&version=v1.2.1&utm_product_name=IntelliJ+IDEA&utm_product_version=2021.1.3&start_row=30&start_col=4&end_row=30&end_col=4) file of our service:
> 
> 1- Create a new grpc service using `NewServiceWithOptions`: [here](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Auri200%2Fmagma.git&branch=state_service_add_print_grpc&file=feg%2Fgateway%2Fservices%2Fs8_proxy%2Fs8_proxy%2Fmain.go&editor=JetBrains&version=v1.2.1&utm_product_name=IntelliJ+IDEA&utm_product_version=2021.1.3&start_row=37&start_col=41&end_row=37&end_col=41) we create a generic `grpc.Server` and preconfigure grpc options like `KeepaliveParams`. We also add`moduleName` and `serviceName` which will be used by controlProxy and Magmad to route those request (the service name matches [service_registry.yml](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Auri200%2Fmagma.git&branch=state_service_add_print_grpc&file=feg%2Fgateway%2Fconfigs%2Fservice_registry.yml&editor=JetBrains&version=v1.2.1&utm_product_name=IntelliJ+IDEA&utm_product_version=2021.1.3&start_row=26&start_col=0&end_row=27&end_col=0) on the feg)
> 
> 2- Parse configuration for that specific service: all services have a [GetConfig](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Auri200%2Fmagma.git&branch=state_service_add_print_grpc&file=feg%2Fgateway%2Fservices%2Fs8_proxy%2Fs8_proxy%2Fmain.go&editor=JetBrains&version=v1.2.1&utm_product_name=IntelliJ+IDEA&utm_product_version=2021.1.3&start_row=37&start_col=41&end_row=37&end_col=41) function which does the task of grabbing the gateway.mconfig and parsing the specific fields for that service. In case of `s8_proxy` we identify the values inside mconfig using `S8ProxyServiceName` [here](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Auri200%2Fmagma.git&branch=state_service_add_print_grpc&file=feg%2Fgateway%2Fservices%2Fs8_proxy%2Fservicers%2Fconfig.go&editor=JetBrains&version=v1.2.1&utm_product_name=IntelliJ+IDEA&utm_product_version=2021.1.3&start_row=39&start_col=7&end_row=39&end_col=7)
> 
> 3- Creation of the implementation of the grpc service: here we create an implementation of `S8ProxyServer` which includes all the grpc calls. That is the actual [s8_proxy](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Auri200%2Fmagma.git&branch=state_service_add_print_grpc&file=feg%2Fgateway%2Fservices%2Fs8_proxy%2Fs8_proxy%2Fmain.go&editor=JetBrains&version=v1.2.1&utm_product_name=IntelliJ+IDEA&utm_product_version=2021.1.3&start_row=40&start_col=33&end_row=40&end_col=33) logic
> 
> 4- Register the implementation of the service into the grpc server
> 
> 5- run the server
> 
> An example of how we parse a config can be found at s6a_proxy [here](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Auri200%2Fmagma.git&branch=state_service_add_print_grpc&file=feg%2Fgateway%2Fservices%2Fs6a_proxy%2Fservicers%2Fconfig.go&editor=JetBrains&version=v1.2.1&utm_product_name=IntelliJ+IDEA&utm_product_version=2021.1.3&start_row=47&start_col=17&end_row=47&end_col=17)
> 
> There we parse a specific key for that service on mconfig file. Once it is parsed we manually put each field on the servcie config file (in case of s6a_proxy this is `S6aProxyConfig`).
> 
> In case mconfig file is not found, we build a default config file using local envars [here](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Auri200%2Fmagma.git&branch=state_service_add_print_grpc&file=feg%2Fgateway%2Fservices%2Fs6a_proxy%2Fservicers%2Fconfig.go&editor=JetBrains&version=v1.2.1&utm_product_name=IntelliJ+IDEA&utm_product_version=2021.1.3&start_row=59&start_col=13&end_row=59&end_col=13)

thanks for these pointers. once our dependency injection story is figured out, will revisit how it meshes with existing service framework stuff.

definitely want to create one easy way to do things for everything going forward, regardless of where the code is running

File: lte/protos/mconfig/mconfigs.proto (hunk 8 above)
Comment (ulaskozat) [reply lte/protos/mconfig/mconfigs.proto:489]: what does accessd do in orc8r? we need a common name that terminates the procedures from (radio) access network rather than mme which is currently mme+amf+sgw-c. I think accessd is a good name to capture a converged access service.

**Review by ulaskozat — APPROVED** [review]
LGTM

File: lte/protos/mconfig/mconfigs.proto (hunk 8 above)
Comment (ardzoht) [reply lte/protos/mconfig/mconfigs.proto:489]: I could be wrong but I think accessd orc8r service is the access manager for operators cc @themarwhal 
 (permissions, certificates) that are tied to a network. I'm not sure if having both `accessd` service names in both orc8r and agw scopes could lead to confusion, but I would advocate to having a different name for the AGW service.

File: lte/protos/mconfig/mconfigs.proto [hunk 10]
```diff
@@ -468,3 +468,22 @@ message GatewayDNSConfigRecordsItems {
   repeated string cname_record = 3;
   string domain = 4;
 }
+
+//----------------------------------------------------------------------------
+// AgwD configs
+//----------------------------------------------------------------------------
+message AgwD {
+  enum LogLevel {
```
Comment (ardzoht) [inline lte/protos/mconfig/mconfigs.proto:476]: there's already a `LogLevel` on `magma/orc8r/protos/common.proto`, but that one does not contain UNSET value, any reason apart from that to not use the common version?

**Review by uri200 — APPROVED** [review]

**Review by ardzoht — APPROVED** [review]

File: lte/protos/mconfig/mconfigs.proto (hunk 8 above)
Comment (themarwhal) [reply lte/protos/mconfig/mconfigs.proto:489]: Yes `accessd` on the orc8r manages access control on Orc8r API access. (like which networks / gateways it has permision to read/write, etc.)

File: lte/protos/mconfig/mconfigs.proto (hunk 10 above)
Comment (themarwhal) [reply lte/protos/mconfig/mconfigs.proto:476]: Having UNSET is probably a good idea as otherwise we can't tell the difference between DEBUG and the value being empty. (Or can we?) I think we want the default value if not specified to be INFO, probably. But I suppose it is confusing that it digerges from the other protos.

Comment (markjen) [reply lte/protos/mconfig/mconfigs.proto:476]: Note: in proto3 all enums should have a 0 = UNSET/INVALID value. Using the 0 value as default is not recommended because 1. distinguishing between unset vs. set can be important sometimes, 2. the default value might change and you can't change explicitly set serialized values.

In this particular case, the common proto also defines a FATAL level, which we no longer support.

**Comment by markjen** [comment]
Note: codecov-action v2 is currently broken https://github.com/codecov/codecov-action/issues/525

**Comment by markjen** [comment]
merging while not all checks are successful; notes:

- `docs-workflow / markdown lint job` has been failing for a few weeks, see #9140 
- `Jenkins` and `Jenkins LTE` tests have been failing for a few months
- `Magma-OAI-Jenkins` has been stuck for 12+ hours, OAI has been pinged
